### PR TITLE
258: Embed videos

### PR DESCRIPTION
**Pull request by f1sh1918**
### Short description

With this feature media via iframe like vimeo videos (and youtube if source is enabled) can be embedded.
I decided to split up the prs

### Proposed changes

- [ ]  implement native with opt-in and source permissions in settings #2602
- [ ] implement web with opt-in and source permissions with cookies and manage it on separate route (external source) this page should be linked on data privacy page

### Side effects

- hopefully none, please test properly all links and features of the webview we are using

### Resolved issues

Fixes: #1213

### CODE REVIEW

File: native/src/utils/renderHtml.ts
Comment (f1sh1918) [inline native/src/utils/renderHtml.ts:97]: just added this. rest was not correct formatted

File: native/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline native/src/components/RemoteContent.tsx:81]:
```suggestion
      if (buildConfig().allowedIframeSources.some(source => event.url.indexOf(source) > 0)) {
```

File: native/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline native/src/components/RemoteContent.tsx:80]: Optional: I think the comment is not really needed as the variable names are pretty imo.
```suggestion
```

File: native/src/utils/renderHtml.ts
Comment (steffenkleinle) [inline native/src/utils/renderHtml.ts:8]: Please revert the formatting, now there are 4 spaces instead of our default of 2. Imo the previous was correct. I think this new indentation is just coming from the fact that you added brackets around the string/didn't use direct return.

File: release-notes/unreleased/258-embed-video.yml
Comment (steffenkleinle) [inline release-notes/unreleased/258-embed-video.yml:8]:
```suggestion
en: Embedding of videos is now supported
de: Einbinden von Videos wird jetzt unterstützt
```

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:17]: Please readd

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:156]: You already using `sandBoxRef.current` above without a non-null check, so it should be fine here as well.
```suggestion
      if (buildConfig().allowedIframeSources.some(source => node.src.indexOf(source) > 0)) {
```

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:158]: Is this a convention that the do not track parameter is called `dnt` or is this just the case for vimeo? Because otherwise we need to keep that in mind for other video platforms.

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:157]:
```suggestion
        const url = new URL(node.src)
```

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:153]: Optional:
```suggestion
    // Remove disallowed iframes from DOM and add do not track parameter
```

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:160]:
```suggestion
      } else {
```

File: native/src/utils/renderHtml.ts
Comment (steffenkleinle) [inline native/src/utils/renderHtml.ts:112]: Question (because I am actually not sure whether it would be easily possible or even if it would be possible if it would make sense): Could we directly strip the `<iframe>` tags from the html in the api-client already?

Then we would only need to do this in one place (but probably for all endpoints with html). So probably it would still not make sense.

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
Tested on android, works fine. Good job!

I can't open any links anymore though (neither deeplinks nor internal ones), without the check it still works
 ``` ts
if (event.navigationType === 'click') {
```

Also, we should take care of what @svenseeberg said [here](https://github.com/digitalfabrik/integreat-app/issues/1213#issuecomment-1822845297):
> Just a note: we need an active opt in of users before content from Vimeo or other services can be loaded.

**Comment by f1sh1918** [comment]
> Tested on android, works fine. Good job!
> 
> I can't open any links anymore though (neither deeplinks nor internal ones), without the check it still works
> 
> ```ts
> if (event.navigationType === 'click') {
> ```
> 
> Also, we should take care of what @svenseeberg said [here](https://github.com/digitalfabrik/integreat-app/issues/1213#issuecomment-1822845297):
> 
> > Just a note: we need an active opt in of users before content from Vimeo or other services can be loaded.

Alright gonna check that. The problem here is that iframe src will also be handled which shouldn't since they are no click event and will be opened in external browser which shouldn't happen.

File: native/src/utils/renderHtml.ts
Comment (f1sh1918) [reply native/src/utils/renderHtml.ts:10]: @steffenkleinle can you specify which string you mean? :)

File: web/src/components/RemoteContent.tsx
Comment (f1sh1918) [reply web/src/components/RemoteContent.tsx:160]: had to declare a const for the `sandBofRef.current` to avoid this check

File: native/src/utils/renderHtml.ts
Comment (f1sh1918) [reply native/src/utils/renderHtml.ts]: I mean we could add this `do not track` in the `api-client` but the rest of the `iframe` handling is different on `native` and `web` so even we need a separate handling.
Additionally i think we not just want to strip this but show a proper message that it was blocked to show the maintainer of the page that he/she is not allowed to use this source instead of just stripping it out.

File: web/src/components/RemoteContent.tsx
Comment (f1sh1918) [reply web/src/components/RemoteContent.tsx:158]: you are right this only works for vimeo, gonna add a check

File: native/src/utils/renderHtml.ts
Comment (LeandraH) [inline native/src/utils/renderHtml.ts:19]:
```suggestion
  (function catchErrors() {
```

Comment (LeandraH) [reply native/src/utils/renderHtml.ts:19]: I think we can remove the comment above now

**Review by LeandraH — APPROVED** [review]
Looks great, thanks! Tested in Firefox and Safari

File: build-configs/BuildConfigType.ts
Comment (steffenkleinle) [inline build-configs/BuildConfigType.ts:54]: I think we should avoid this term if possible bc of discrimination, why did you change it from `allowed`?

File: native/src/utils/renderHtml.ts
Comment (steffenkleinle) [inline native/src/utils/renderHtml.ts:97]: Please use speaking names:
```suggestion
    iframes.forEach((iframe) => {
```

File: web/src/utils/doNotTrack.ts
Comment (steffenkleinle) [inline web/src/utils/doNotTrack.ts:3]:
```suggestion
  if (iframe.src.includes('vimeo')) {
```

File: web/src/utils/doNotTrack.ts
Comment (steffenkleinle) [inline web/src/utils/doNotTrack.ts:1]: I think this is self explanatory, isn't it? :see_no_evil:

Comment (steffenkleinle) [reply web/src/utils/doNotTrack.ts:1]: Does it make sense to move this to the api-client as this is probably the same logic for native?

File: web/src/components/RemoteContent.tsx
Comment (steffenkleinle) [inline web/src/components/RemoteContent.tsx:159]:
```suggestion
      if (buildConfig().whiteListedIframeSources.some(iframe.src.includes)) {
```

**Review by steffenkleinle — COMMENTED** [review]
There is a few changes for native here, will you wait for native to merge this? Should we review those changes here? I think it would be better to actually split them.

File: web/src/utils/doNotTrack.ts
Comment (f1sh1918) [reply web/src/utils/doNotTrack.ts:1]: In general this makes sense, but the problem is we can not import function in the string template. Or lets say import functions and pass them parameters. I tried this several time with the `t()` to pass sources to the translations

**Comment by f1sh1918** [comment]
> There is a few changes for native here, will you wait for native to merge this? Should we review those changes here? I think it would be better to actually split them.

Yes your are right sorry for the confusion. I'm gonna create two separate branches and pr and integrate your suggestions

**Pull request closed without merging**
